NTRDMA: Resolving issue with 40ms TCP latency

saman-eitr opened the pull request:
This patch fixes the issue with constant 40ms response latency, in special communication cases. Prior to this change, some applications may suffer from huge latency and their poll completion may stuck for 40ms. By enabling TCP_NODELAY flag, we can make sure the TCP protocol does not introduce any artificial waiting time during the packet transmission.

Signed-off by: Saman Biookaghazadeh saman@eitrsystems.com

saman-eitr committed:
NTC: ntc_tcp: set TCP_NODELAY in ntc_tcp

This patch fixes the issue with 40ms in ntc_tcp.  Prior to this change,
applications may see work completions delayed by about 40ms.  By
enabling TCP_NODELAY flag, we can request that the TCP protocol not
introduce any extra delay during the transmission.

Signed-off-by: Saman Biookaghazadeh <saman@eitrsystems.com>
Signed-off-by: Allen Hubbe <Allen.Hubbe@emc.com>

allenbh commented on drivers/ntc/ntc_tcp.c:
```diff
@@ -827,5 +845,6 @@ static void __exit ntc_tcp_exit(void)
 }
-module_exit(ntc_tcp_exit);
 
+
```
From checkpatch.pl:

``` txt
CHECK: Please don't use multiple blank lines
#115: FILE: drivers/ntc/ntc_tcp.c:849:

+
```

allenbh replied:
Better to just leave module_exit() where it was.  It has nothing to do with the rest of the patch.

saman-eitr replied:
Done!

allenbh commented:
The end result looks good, but this should be one commit.  Let's squash these into one good commit.

``` sh
git rebase -i
```

You should see this:

``` txt
pick b567b318c042 NTRDMA: Resolving issue with 40ms TCP latency\n\nThis ...
pick bff9c9661b84 -
pick 40035ff3076e NTRDMA: Resolving issue with 40ms TCP latency
pick ac799d0ff554 NTRDMA: Resolving issue with 40ms TCP latency
pick 5a60e2ba699e NTRDMA: Resolving issue with 40ms TCP latency
pick 4743d2f41031 NTRDMA: Resolving issue with 40ms TCP latency
pick ea17666eaf2f NTRDMA: Resolving issue with 40ms TCP latency
```

Change it to this (all but the first one say "squash"), then save and exit your editor:

``` txt
pick b567b318c042 NTRDMA: Resolving issue with 40ms TCP latency\n\nThis ...
squash bff9c9661b84 -
squash 40035ff3076e NTRDMA: Resolving issue with 40ms TCP latency
squash ac799d0ff554 NTRDMA: Resolving issue with 40ms TCP latency
squash 5a60e2ba699e NTRDMA: Resolving issue with 40ms TCP latency
squash 4743d2f41031 NTRDMA: Resolving issue with 40ms TCP latency
squash ea17666eaf2f NTRDMA: Resolving issue with 40ms TCP latency
```

Finally, you will be presented with the squashed commit message.  Choose the one that looks best, and delete all the rest.  Make any final edits to the message, then save and exit your editor.

All the changes should now be in one commit.  Push your branch (you will need force because of rebase).

``` sh
git push -f
```

saman-eitr commented:
Done with the rebase!

allenbh commented:
Thanks for the patch!  Tomorrow I will test, and assuming all goes well, I will accept these changes.

I will take care of wrapping the commit message tomorrow when I accept the changes:

``` txt
WARNING: Possible unwrapped commit description (prefer a maximum 75 chars per line)
#6: 
This patch fixes the issue with constant 40ms response latency, in special communication cases. Prior to this change, some applications may suffer from huge latency and their poll completion may stuck for 40ms. By enabling TCP_NODELAY flag, we can make sure the TCP protocol does not introduce any artificial waiting time during the packet transmission.

total: 0 errors, 1 warnings, 0 checks, 62 lines checked
```

saman-eitr commented:
Thanks much for the help and patience.

allenbh commented:
Do you have a preference for which of your identities is associated with the commit?

``` txt
git show --pretty=full | head
commit f05b0a5de252b392565e63575690295a5b1fbfa2
Author: Saman Biookaghazadeh <...@gmail...>
Commit: Saman Biookaghazadeh <...@eitr...>
```

saman-eitr commented:
For this commit I'm gonna go with the eitrsystems id, since the patch has been created during my existence here.

allenbh commented:
I wonder if I applied this right.  Or, I wonder if due to the number of small messages added to the ntc_tcp stream, allowing the tcp stack to buffer them is more efficient most of the time, but the tcp delay is noticeable when there are very few small outstanding operations, not enough to fill the send buffer.

``` txt
ibv_rc_pingpong peer
  local address:  LID 0xbeef, QPN 0x000000, PSN 0x589d8e, GID ::
  remote address: LID 0xbeef, QPN 0x000000, PSN 0x402ff2, GID ::
8192000 bytes in 105.54 seconds = 0.62 Mbit/sec
1000 iters in 105.54 seconds = 105536.41 usec/iter
```

Can you add a module parameter to enable and disable tcp nodelay, and show results for various tests like pingpong, qperf, and the test where you noticed the 40ms delay?  I'm interested to compare the results with and without nodelay.

I have tentatively accepted this change into the [proposed updates](/ntrdma/ntrdma/commits/pu) branch as 100fd9ee43a120eb414c945f9383d45d326ea78d.  Thanks for the patch!

Commit:
NTC: ntc_tcp: add module parameter nodelay

Add module parameter to enable or disable TCP_NODELAY, with default set
to enable.

Signed-off-by: Allen Hubbe <Allen.Hubbe@emc.com>

allenbh commented:
Please ignore all the "referenced this pull request" garbage.  I merged content from Linux 4.7.  Some of the commits coincidentally have `#2` somewhere in the commit message, which github interprets as #2.

allenbh commented:
Thanks! Merged.
